### PR TITLE
COM-10931: Precisions for statuses (names, groups, order, etc.)

### PR DESCRIPTION
**Pull request by dememax**
If I understand correctly, we have groups of statuses in strict order, and it is impossible to mix these groups up (for example, the usage of `has_pending` in the features cycle of the sync command).

Questions based on the existing algorithm in `GUW._sync_at()`:
1. `integrated` cannot appear between `merging` и `pending` (protected by `has_pending`); but what about `merged`, why there is no check for this?
1. Can we mix up `merging` и `pending` - there is no protection, there is no difference in treating them at all (it looks like, the only difference is to have the pr field or not).
1. `merging` could be multiple? Or just one at time?
1. What is the difference between `prev_feature` and `prev_active_feature`?

If so, we should change the algo; this is not an appropriate approach in `GUW._sync_at()`:
```Python
    for i in features:
        if i.status == integrated:
        elif i.status == merged:
        elif i.status == merging or pending:
        else # wrong name; I've added this in the PR
```
to
1. get group indexes using finding first in features (find first `integrated`, then first `merged`, and so on)
1. check indexes
    1. `integrated` - if any, comes first
    1. `merged` - if any, comes second
    1. `merging` - if any, comes third (and logically could be only one?)
    1. `pending`  - final group
1. ignore `integrated` and `merged`, only reporting, taking the latest as a base for rebase
1. rebasing `merging` and `pending`
 
Maybe we can change the syntax of the config file, as well, to have only three groups inside features (if my assumption on the difference between `merging` and `pending` is right).
```TOML
# remotes, target, source stuff as is

[[features.integrated]]
remote = "origin"
name = "feature1"
pr = "https://github/fluendo/git-upstream-workflow/pull-requests/XXX"

[[features.integrated]]
remote = "origin"
name = "another already integrated feature"
pr = "https://github/fluendo/git-upstream-workflow/pull-requests/XXX"

[[features.merged]]
remote = "origin"
name = "feature2"
pr = "https://github/fluendo/git-upstream-workflow/pull-requests/XXX"

[[features.pending]]
remote = "origin"
name = "currently merging feature"
pr = "https://github/fluendo/git-upstream-workflow/pull-requests/XXX"

[[features.pending]]
remote = "origin"
name = "still pending feature"
status = "pending"
```

### CODE REVIEW

**Comment by dememax** [comment]
Our rules for commits are not the same as the rules in the community:
we usually put the jira item reference.
So, by using guw utility we should follow the community notation.

**Comment by dememax** [comment]
See also signaled problems on Slack:
https://fluendo.slack.com/archives/C2VDNCG69/p1721903938597739

**Comment by turran** [comment]
Hello @dememax,

I think there is a misunderstanding of what each feature is. A feature is not a commit; it is a branch, so it has a commit or multiple commits on it. The order is important and has a meaning on the corresponding commits of each branch. If the features are:

```
[[ features ]]
name = A
[[ features ]]
name = B
```
It means that B has its own commits plus commits from A.  So, you can never merge B, without merging A. This workflow is a rebase based workflow, not a merge based one.

Answering your questions:

>   integrated cannot appear between merging и pending (protected by has_pending); but what about merged, why there is no check for this?

In theory, there can't be a higher-order status on the next feature. If feature A is on merging, B can not be `merging` or anything after it: `merged`, `integrated`.

>  Can we mix up merging и pending - there is no protection, there is no difference in treating them at all (it looks like, the only difference is to have the pr field or not).

Correct, `merging` and `pending` difference is just a semantic difference on the .toml, but not a technical limitation from the POV of git commands

>  merging could be multiple? Or just one at time?

this is related to the concepts I commented at the beginning. It can only be one. 

>  What is the difference between prev_feature and prev_active_feature?

It is due to how to rebase, once  a feature is `integrated`, you discard such branch to rebase from it.

**Comment by dememax** [comment]
> Hello @dememax,
> 
> I think there is a misunderstanding of what each feature is. A feature is not a commit; it is a branch, so it has a commit or multiple commits on it. The order is important and has a meaning on the corresponding commits of each branch. If the features are:
> 
> ```
> [[ features ]]
> name = A
> [[ features ]]
> name = B
> ```
> It means that B has its own commits plus commits from A.  So, you can never merge B, without merging A. This workflow is a rebase based workflow, not a merge based one.

No, I understand that it's about branches, not commits.

Another interesting observation: pointing the `remote` in `features` is useless in this case, it is already in the `source`, and cannot be changed on the `features` level.

File: guw/main.py
Comment (turran) [inline guw/main.py]: Update the help message to be just "working folder" instead of "temporary" as it is no longer the case with the new optoin

**Review by turran — CHANGES_REQUESTED** [review]

File: guw/main.py
Comment (dememax) [reply guw/main.py]: Done. + little improvement.

Please, re-review. Thanks!

File: guw/main.py
Comment (turran) [inline guw/main.py:91]: The remote is no longer included in the checkout. I prefer to keep it explicit otherwise, it can collision with other branches

**Review by turran — CHANGES_REQUESTED** [review]

File: guw/main.py
Comment (dememax) [reply guw/main.py:91]: @turran Are you sure?
Please, rereview more attentively and reread my commit message!
I haven't changed anything in the logic. Only:

1. `target_branch` was never used. Deleted.
2. Independently of the previous, all complex expressions `self.config['target']['branch']` (or with double quotes) are substituted with variable that hold this expression.

Comment (turran) [reply guw/main.py:91]: I see. Good to go then. I still want to keep an eye on the explicit checkout, not because of your commit but because of the current logic.

**Review by turran — APPROVED** [review]

**Pull request merged**
